### PR TITLE
[CELEBORN-558] Bump Ratis to 2.5.1 and fix API changes

### PR DESCRIPTION
**Pull request by kaijchen**
### What changes were proposed in this pull request?

Bump Ratis to 2.5.0 and fix API changes.

### Why are the changes needed?

Ratis 2.5.0 has been released and brings several improvements and bug fixes.
However, a few API changes need to be fixed due to celeborn is using some internal APIs of Ratis.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing CI.

### CODE REVIEW

File: master/src/main/java/org/apache/ratis/statemachine/impl/SimpleStateMachineStorageUtil.java
Comment (AngersZhuuuu) [inline master/src/main/java/org/apache/ratis/statemachine/impl/SimpleStateMachineStorageUtil.java:23]: Seems unnecessary?

Comment (kaijchen) [reply master/src/main/java/org/apache/ratis/statemachine/impl/SimpleStateMachineStorageUtil.java:23]: Turning package private methods to public.

**Review by FMX — APPROVED** [review]
LGTM. I think this patch will need some tests before merging. I'll test it later.

**Review by AngersZhuuuu — APPROVED** [review]
LGTM

**Comment by waitinfuture** [comment]
> 

Thanks @FMX

**Comment by kaijchen** [comment]
Thanks @FMX and @AngersZhuuuu for reviewing. Ratis 2.5.1 is currently under voting. We can wait for its release and upgrade to 2.5.1 directly.

**Comment by FMX** [comment]
> Thanks @FMX and @AngersZhuuuu for reviewing. Ratis 2.5.1 is currently under voting. We can wait for its release and upgrade to 2.5.1 directly.

Thanks, I'll test this pr after you upgrade ratis to 2.5.1.

**Comment by kaijchen** [comment]
> Thanks, I'll test this pr after you upgrade ratis to 2.5.1.

Done.

**Comment by FMX** [comment]
Thanks, I'll test this patch today.

**Comment by FMX** [comment]
After I killed a node and waited for cache timeout, then I got lots of logs as below:
```
23/05/09 14:56:59,542 INFO [3@group-47BEDE733167-FollowerState] RaftServerConfigKeys: raft.server.rpc.first-election.timeout.min = 3s (fallback to raft.server.rpc.timeout.min)
23/05/09 14:57:04,172 INFO [3@group-47BEDE733167-FollowerState] RaftServerConfigKeys: raft.server.rpc.first-election.timeout.min = 3s (fallback to raft.server.rpc.timeout.min)
23/05/09 14:57:08,210 INFO [3@group-47BEDE733167-FollowerState] RaftServerConfigKeys: raft.server.rpc.first-election.timeout.min = 3s (fallback to raft.server.rpc.timeout.min)
23/05/09 14:57:12,787 INFO [3@group-47BEDE733167-FollowerState] RaftServerConfigKeys: raft.server.rpc.first-election.timeout.min = 3s (fallback to raft.server.rpc.timeout.min)
23/05/09 14:57:16,977 INFO [3@group-47BEDE733167-FollowerState] RaftServerConfigKeys: raft.server.rpc.first-election.timeout.min = 3s (fallback to raft.server.rpc.timeout.min)
```

This won't happen when using ratis 2.4.0. So I think that there might be something wrong with ratis 2.5.1. @kaijchen

**Review by FMX — CHANGES_REQUESTED** [review]
There is a question to be solved before this patch merge.

**Comment by kaijchen** [comment]
> This won't happen when using ratis 2.4.0. So I think that there might be something wrong with ratis 2.5.1

Is `raft.server.rpc.first-election.timeout.min` configured in your environment?

**Comment by FMX** [comment]
> 

This config is not configured in Celeborn. Is it needed to be set explicitly?

**Comment by pan3793** [comment]
Looks like a noisy log, maybe it's better to log one time or simply change the level to debug @kaijchen

**Comment by pan3793** [comment]
@FMX as we are here, I remember there were troubles when using Ratis w/ gRPC on K8s, do you have a chance to test it again w/ this brand-new version?

**Comment by FMX** [comment]
Sounds reasonable, I’ll run some tests.

在 2023年5月9日 17:59:36 上，Cheng Pan ***@***.***> 写道：

**Comment by AngersZhuuuu** [comment]
> After I killed a node and waited for cache timeout, then I got lots of logs as below:
> 
> ```
> 23/05/09 14:56:59,542 INFO [3@group-47BEDE733167-FollowerState] RaftServerConfigKeys: raft.server.rpc.first-election.timeout.min = 3s (fallback to raft.server.rpc.timeout.min)
> 23/05/09 14:57:04,172 INFO [3@group-47BEDE733167-FollowerState] RaftServerConfigKeys: raft.server.rpc.first-election.timeout.min = 3s (fallback to raft.server.rpc.timeout.min)
> 23/05/09 14:57:08,210 INFO [3@group-47BEDE733167-FollowerState] RaftServerConfigKeys: raft.server.rpc.first-election.timeout.min = 3s (fallback to raft.server.rpc.timeout.min)
> 23/05/09 14:57:12,787 INFO [3@group-47BEDE733167-FollowerState] RaftServerConfigKeys: raft.server.rpc.first-election.timeout.min = 3s (fallback to raft.server.rpc.timeout.min)
> 23/05/09 14:57:16,977 INFO [3@group-47BEDE733167-FollowerState] RaftServerConfigKeys: raft.server.rpc.first-election.timeout.min = 3s (fallback to raft.server.rpc.timeout.min)
> ```
> 
> This won't happen when using ratis 2.4.0. So I think that there might be something wrong with ratis 2.5.1. @kaijchen

Checked the code, if want print this, we may need to add a default value for this key. Not a problem

**Review by pan3793 — APPROVED** [review]

**Review by AngersZhuuuu — APPROVED** [review]
LGTM, WDYT @FMX

**Comment by FMX** [comment]
> LGTM, WDYT @FMX
I see your comments. I think this PR is fine but the logs will continue to write after a node is restarted.

If adding a config can solve this, the config should be added.  @kaijchen

**Review by FMX — APPROVED** [review]
The code looks fine to me. I'll handle the logs later.

**Pull request merged**
